The `generate` command crashes for `hangouts` service

On current master:

```
$ go run cli/main.go generate hangouts
Generating URL for hangouts using basic generator
Enter the configuration values as prompted

panic: interface conversion: *hangouts.Config is not types.ServiceConfig: missing method GetURL

goroutine 1 [running]:
github.com/containrrr/shoutrrr/pkg/format.GetServiceConfigFormat(0x169c820, 0xc00000ea60, 0x0, 0x0, 0x0, 0x1, 0x0)
	/Users/jonas/projects/blockdaemon/shoutrrr-upstream/pkg/format/formatter.go:38 +0x17e
github.com/containrrr/shoutrrr/pkg/generators/basic.(*Generator).Generate(0x19b14a8, 0x169c820, 0xc00000ea60, 0xc0001a0120, 0xc000075590, 0x1, 0x1, 0x0, 0xc000199cc8, 0x13f6a86, ...)
	/Users/jonas/projects/blockdaemon/shoutrrr-upstream/pkg/generators/basic/basic.go:23 +0xe5
github.com/containrrr/shoutrrr/cli/cmd/generate.Run(0x1977240, 0xc000075590, 0x1, 0x1)
	/Users/jonas/projects/blockdaemon/shoutrrr-upstream/cli/cmd/generate/generate.go:119 +0x774
github.com/spf13/cobra.(*Command).execute(0x1977240, 0xc000075570, 0x1, 0x1, 0x1977240, 0xc000075570)
	/Users/jonas/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:842 +0x2c2
github.com/spf13/cobra.(*Command).ExecuteC(0x1977a20, 0x0, 0x0, 0xc000096058)
	/Users/jonas/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:943 +0x336
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/jonas/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:883
main.main()
	/Users/jonas/projects/blockdaemon/shoutrrr-upstream/cli/main.go:26 +0x2d
exit status 2
```

--- a/pkg/services/hangouts/hangouts.go
+++ b/pkg/services/hangouts/hangouts.go
@@ -40,8 +40,10 @@ func (service *Service) Send(message string, _ *types.Params) error {
 		return err
 	}
 
+	postURL := getAPIURL(config)
+
 	jsonBuffer := bytes.NewBuffer(jsonBody)
-	resp, err := http.Post(config.URL.String(), "application/json", jsonBuffer)
+	resp, err := http.Post(postURL.String(), "application/json", jsonBuffer)
 
 	if err != nil {
 		return fmt.Errorf("failed to send notification to Hangouts Chat: %s", err)
@@ -54,4 +56,17 @@ func (service *Service) Send(message string, _ *types.Params) error {
 	}
 
 	return nil
+}
+
+func getAPIURL(config *Config) *url.URL {
+	query := url.Values{}
+	query.Set("key", config.Key)
+	query.Set("token", config.Token)
+
+	return &url.URL{
+		Path:     config.Path,
+		Host:     config.Host,
+		Scheme:   "https",
+		RawQuery: query.Encode(),
+	}
 }

--- a/pkg/services/hangouts/hangouts_config.go
+++ b/pkg/services/hangouts/hangouts_config.go
@@ -1,23 +1,66 @@
 package hangouts
 
 import (
+	"errors"
 	"net/url"
 
+	"github.com/containrrr/shoutrrr/pkg/format"
 	"github.com/containrrr/shoutrrr/pkg/services/standard"
+	"github.com/containrrr/shoutrrr/pkg/types"
 )
 
 // Config for use within the Hangouts Chat plugin.
 type Config struct {
 	standard.EnumlessConfig
-	URL *url.URL
+	Host  string `default:"chat.googleapis.com"`
+	Path  string
+	Token string
+	Key   string
+}
+
+// GetURL returns a URL representation of it's current field values
+func (config *Config) GetURL() *url.URL {
+	resolver := format.NewPropKeyResolver(config)
+	return config.getURL(&resolver)
+}
+
+// SetURL updates a ServiceConfig from a URL representation of it's field values
+func (config *Config) SetURL(url *url.URL) error {
+	resolver := format.NewPropKeyResolver(config)
+	return config.setURL(&resolver, url)
 }
 
 // SetURL updates a ServiceConfig from a URL representation of it's field values.
-func (config *Config) SetURL(url *url.URL) error {
-	config.URL = url
-	config.URL.Scheme = "https"
+func (config *Config) setURL(_ types.ConfigQueryResolver, serviceURL *url.URL) error {
+	config.Host = serviceURL.Host
+	config.Path = serviceURL.Path
+
+	query := serviceURL.Query()
+	config.Key = query.Get("key")
+	config.Token = query.Get("token")
+
+	if config.Key == "" {
+		return errors.New("missing field 'key'")
+	}
+
+	if config.Key == "" {
+		return errors.New("missing field 'token'")
+	}
 
 	return nil
+}
+
+func (config *Config) getURL(_ types.ConfigQueryResolver) *url.URL {
+	query := url.Values{}
+	query.Set("key", config.Key)
+	query.Set("token", config.Token)
+
+	return &url.URL{
+		Host:     config.Host,
+		Path:     config.Path,
+		RawQuery: query.Encode(),
+		Scheme:   Scheme,
+	}
 }
 
 const (

--- a/pkg/services/hangouts/hangouts_test.go
+++ b/pkg/services/hangouts/hangouts_test.go
@@ -21,6 +21,6 @@ var _ = Describe("the Hangouts Chat plugin URL building", func() {
 		config.SetURL(configURL)
 
 		expectedURL := "https://chat.googleapis.com/v1/spaces/FOO/messages?key=bar&token=baz"
-		Expect(config.URL.String()).To(Equal(expectedURL))
+		Expect(getAPIURL(&config).String()).To(Equal(expectedURL))
 	})
 })

--- a/pkg/services/smtp/smtp_encmethod.go
+++ b/pkg/services/smtp/smtp_encmethod.go
@@ -52,4 +52,5 @@ func useImplicitTLS(encryption encMethod, port uint16) bool {
 	}
 }
 
+// ImplicitTLSPort is de facto standard SMTPS port
 const ImplicitTLSPort = 465

--- a/pkg/services/telegram/telegram_json.go
+++ b/pkg/services/telegram/telegram_json.go
@@ -1,6 +1,6 @@
 package telegram
 
-// JSON to be used as a notification payload for the telegram notification service
+// SendMessagePayload is the notification payload for the telegram notification service
 type SendMessagePayload struct {
 	Text                string `json:"text"`
 	ID                  string `json:"chat_id"`

--- a/pkg/services/zulip/zulip.go
+++ b/pkg/services/zulip/zulip.go
@@ -59,7 +59,7 @@ func (service *Service) Initialize(configURL *url.URL, logger *log.Logger) error
 	service.Logger.SetLogger(logger)
 	service.config = &Config{}
 
-	if err := service.config.SetURL(nil, configURL); err != nil {
+	if err := service.config.setURL(nil, configURL); err != nil {
 		return err
 	}
 
@@ -67,24 +67,26 @@ func (service *Service) Initialize(configURL *url.URL, logger *log.Logger) error
 }
 
 func (service *Service) doSend(config *Config, message string) error {
-	apiURL := service.getURL(config)
+	apiURL := service.getAPIURL(config)
 	payload := CreatePayload(config, message)
 	res, err := http.Post(apiURL, "application/x-www-form-urlencoded", strings.NewReader(payload.Encode()))
 
-	if res.StatusCode != http.StatusOK {
-		return fmt.Errorf("failed to send notification to service, response status code %s", res.Status)
+	if err == nil && res.StatusCode != http.StatusOK {
+		err = fmt.Errorf("response status code %s", res.Status)
 	}
 
-	return err
+	if err != nil {
+		return fmt.Errorf("failed to send zulip message: %s", err)
+	}
+
+	return nil
 }
 
-func (service *Service) getURL(config *Config) string {
-	url := url.URL{
+func (service *Service) getAPIURL(config *Config) string {
+	return (&url.URL{
 		User:   url.UserPassword(config.BotMail, config.BotKey),
 		Host:   config.Host,
-		Path:   config.Path,
+		Path:   "api/v1/messages",
 		Scheme: "https",
-	}
-
-	return url.String()
+	}).String()
 }

--- a/pkg/services/zulip/zulip_config.go
+++ b/pkg/services/zulip/zulip_config.go
@@ -2,22 +2,35 @@ package zulip
 
 import (
 	"errors"
+	"github.com/containrrr/shoutrrr/pkg/format"
+	"github.com/containrrr/shoutrrr/pkg/services/standard"
 	"github.com/containrrr/shoutrrr/pkg/types"
 	"net/url"
 )
 
 // Config for the zulip service
 type Config struct {
-	BotMail string
-	BotKey  string
-	Host    string
-	Path    string
-	Stream  string `key:"stream"`
-	Topic   string `key:"topic"`
+	standard.EnumlessConfig
+	BotMail string `desc:"Bot e-mail address"`
+	BotKey  string `desc:"API Key"`
+	Host    string `desc:"API server hostname"`
+	Stream  string `key:"stream" description:"Target stream name"`
+	Topic   string `key:"topic,title" default:""`
 }
 
 // GetURL returns a URL representation of it's current field values
-func (config *Config) GetURL(_ types.ConfigQueryResolver) *url.URL {
+func (config *Config) GetURL() *url.URL {
+	resolver := format.NewPropKeyResolver(config)
+	return config.getURL(&resolver)
+}
+
+// SetURL updates a ServiceConfig from a URL representation of it's field values
+func (config *Config) SetURL(url *url.URL) error {
+	resolver := format.NewPropKeyResolver(config)
+	return config.setURL(&resolver, url)
+}
+
+func (config *Config) getURL(_ types.ConfigQueryResolver) *url.URL {
 	query := &url.Values{}
 
 	if config.Stream != "" {
@@ -31,14 +44,13 @@ func (config *Config) GetURL(_ types.ConfigQueryResolver) *url.URL {
 	return &url.URL{
 		User:     url.UserPassword(config.BotMail, config.BotKey),
 		Host:     config.Host,
-		Path:     config.Path,
 		RawQuery: query.Encode(),
 		Scheme:   Scheme,
 	}
 }
 
 // SetURL updates a ServiceConfig from a URL representation of it's field values
-func (config *Config) SetURL(_ types.ConfigQueryResolver, serviceURL *url.URL) error {
+func (config *Config) setURL(_ types.ConfigQueryResolver, serviceURL *url.URL) error {
 	var ok bool
 
 	config.BotMail = serviceURL.User.Username()
@@ -59,7 +71,6 @@ func (config *Config) SetURL(_ types.ConfigQueryResolver, serviceURL *url.URL) e
 		return errors.New(string(MissingHost))
 	}
 
-	config.Path = "api/v1/messages"
 	config.Stream = serviceURL.Query().Get("stream")
 	config.Topic = serviceURL.Query().Get("topic")
 
@@ -72,7 +83,6 @@ func (config *Config) Clone() *Config {
 		BotMail: config.BotMail,
 		BotKey:  config.BotKey,
 		Host:    config.Host,
-		Path:    config.Path,
 		Stream:  config.Stream,
 		Topic:   config.Topic,
 	}
@@ -86,7 +96,7 @@ const (
 // CreateConfigFromURL to use within the zulip service
 func CreateConfigFromURL(serviceURL *url.URL) (*Config, error) {
 	config := Config{}
-	err := config.SetURL(nil, serviceURL)
+	err := config.setURL(nil, serviceURL)
 
 	return &config, err
 }

--- a/pkg/services/zulip/zulip_test.go
+++ b/pkg/services/zulip/zulip_test.go
@@ -86,7 +86,6 @@ var _ = Describe("the zulip service", func() {
 					BotMail: "bot-name@zulipchat.com",
 					BotKey:  "correcthorsebatterystable",
 					Host:    "example.zulipchat.com",
-					Path:    "api/v1/messages",
 					Stream:  "foo",
 					Topic:   "bar",
 				}
@@ -100,7 +99,6 @@ var _ = Describe("the zulip service", func() {
 					BotMail: "bot-name@zulipchat.com",
 					BotKey:  "correcthorsebatterystable",
 					Host:    "example.zulipchat.com",
-					Path:    "api/v1/messages",
 					Stream:  "foo",
 					Topic:   "bar",
 				}
@@ -121,7 +119,6 @@ var _ = Describe("the zulip service", func() {
 					BotMail: "bot-name@zulipchat.com",
 					BotKey:  "correcthorsebatterystable",
 					Host:    "example.zulipchat.com",
-					Path:    "api/v1/messages",
 					Stream:  "foo",
 					Topic:   "bar",
 				}
@@ -137,7 +134,7 @@ var _ = Describe("the zulip service", func() {
 					Stream:  "foo",
 					Topic:   "bar",
 				}
-				url := config.GetURL(nil)
+				url := config.GetURL()
 				Expect(url.String()).To(Equal("zulip://bot-name%40zulipchat.com:correcthorsebatterystable@example.zulipchat.com?stream=foo&topic=bar"))
 			})
 		})
@@ -149,7 +146,7 @@ var _ = Describe("the zulip service", func() {
 					Host:    "example.zulipchat.com",
 					Stream:  "foo",
 				}
-				url := config.GetURL(nil)
+				url := config.GetURL()
 				Expect(url.String()).To(Equal("zulip://bot-name%40zulipchat.com:correcthorsebatterystable@example.zulipchat.com?stream=foo"))
 			})
 		})

